### ISSUE TITLE
Live Wallpaper for widgets (so can be used on lockscreen)

### ISSUE THREAD

**Issue by ericfont** [forrestguice/SuntimesWidget#513]
I'd really like to have a Suntimes display in my lockscreen.  I don't think the current Android/Lineage version allows regular widgets to display in the lockscreen (I think for security reasons), but live wallpaper can display in the lockscreen as a workaround.  Maybe if simple enough could make a version of each widget as a live wallpaper.  I might even make an attempt coding a wallpaper design that works for me and that avoids overlapping my phone's lockscreen notifications area.

**Comment by forrestguice** [comment]
This seemed like fun and I had some time. There's now a live wallpaper implementation in the NaturalHour [wallpaper branch](https://github.com/forrestguice/NaturalHour/tree/wallpaper).

**Comment by ericfont** [comment]
Great, if you can provide me an apk I'd be happy to test and provide feedback...I don't have my computer setup to build android apps at the moment or else I would myself.

**Comment by forrestguice** [comment]
Sure. I think we can use the debug builds automatically created by the CI.

From https://github.com/forrestguice/NaturalHour/pull/10 click `Show All Checks` -> `Details`, then click `Summary` (left side). There should be a list of 'artifacts' - `app` should be an unsigned debug build.

The wallpaper seems to work well enough, but probably needs options for alignment. Its currently hardcoded to the top of the screen. A smaller version might bounce from one corner of the screen to the other. Lots of possibilities.. Its not exactly the widget you wanted though, but it demonstrates its feasible.

**Comment by ericfont** [comment]
Hey this is great...it works!  There it is on my lockscreen:

![Screenshot_20211102-021739.png](https://user-images.githubusercontent.com/6502474/139795864-4b26cc39-333a-4bf8-8493-21386fd749e7.png)

And yeah it is a bit too busy for a regular wallpaper.  Though if I can get around to it make a minimalistic trimmed down version.

I am thinking that maybe the whole convert widget(s) to live wallpaper thing would be most effectively done by the OS, so that every single app's widgets could be made into a live wallpaper.

**Comment by ericfont** [comment]
Hmm...i woke up ~8 hours layer but my wallpaper reverted back to default.  Maybe some glitch caused it to crash while i was sleeping.

I also notice when rotating the screen, your live wallpaper does reorient itself to point upwards.  However it doesn't resize itself to fit within the screen's new dimensions *until* i press the power button to turn off the screen or trigger a screen redraw.

**Comment by forrestguice** [comment]
> my wallpaper reverted back to default. Maybe some glitch caused it to crash while i was sleeping.

That's a possibility. I've noticed it revert back if it stops running for any reason (crashes). If so there will be something in the logcat. 

Another possibility is battery optimization intentionally killed it. Its technically just another Service which Android can stop at any time, but some phones are worse than others (https://dontkillmyapp.com/). If so maybe white-listing the app from the "Battery optimization" page will keep it running. The actual battery use should be minimal. It only (re)draws every 30s when visible.

>  it doesn't resize itself to fit within the screen's new dimensions
This should be fixed now. Another debug apk is available (same place).

**Comment by ericfont** [comment]
ok thanks, I see my android did have battery optimization enabled for Suntimes, and so I have disabled it.  (I only saw the main app Suntimes listed in the battery optimization page, not the widget apps, though.)

I've downloaded the new apk and see if I have any more crashes and if so will post the logcat.

**Comment by forrestguice** [comment]
Hrm.. those some interesting log messages. I'm not sure they are all related. The first set was generated by your (suntimes) widgets. It looks relatively normal, except those lines "failed to connect to jdwp control socket" - I'm not sure what that's about (or even is).

The second one (crash) is likely the wallpaper (hard to say), but there's nothing there to help us. Instead, try grepping for 'naturalhour'. It owns the live wallpaper process and not suntimes (which doesn't run directly in this case, only its contentprovider). Maybe there is something in the lines preceding the crash.

The same goes for the battery optimization stuff - its "natural hour" should be excluded ("not optimized") vs Suntimes. It doesn't hurt to whitelist Suntimes too though, but the consequences are limited to the alarms (may be delayed up to 15min, or failing to reschedule after a reboot, etc). BTW, I think that's a confusing UI.. it displays the whitelist first, then forces you to select "all apps" and search for the app you want to exclude.

**Comment by ericfont** [comment]
> The same goes for the battery optimization stuff - its "natural hour" should be excluded ("not optimized") vs Suntimes. It doesn't hurt to whitelist Suntimes too though, 

Ah...I see I had earlier just whitelisted (i.e. set to "Not optimized") the Suntimes app but forgot to whitelist Natural Hour.  Now I have whitelisted (i.e. set to "Not optimized") the Natural Hour app too and will restart everything and look out for crashes.

**Comment by forrestguice** [comment]
Yeah.. I'm not sure. Seems like the lines are truncated. Actually, grepping for 'suntimes' should return lines for naturalhour (as well as the other widgets). The full package in question is `com.forrestguice.suntimes.naturalhour`. There should be at least one log message in there; `D/NaturalHourWallpaperEngine: updateViews: -1` every time it draws to the screen.

**Comment by ericfont** [comment]
ok.  Just providing for the record I've rebooted my phone after applying the suntimes and have read logcat and grepped suntimes:

[logcat_reboot_20211103_suntimes.txt](https://github.com/forrestguice/SuntimesWidget/files/7470275/logcat_reboot_20211103_suntimes.txt)

So that will provide me a reference for what the logcat looks like when it was working, that way I have a better idea what is an unusual message.

**Comment by ericfont** [comment]
I just noticed if i have my phone horizontal before i turn the screen on, then the lockscreen has the widget centered to fit inside the screen...which is good...

But if i then unlock my screen and then rotate my phone to be vertical, then the wallpaper doesn't resize immediate and so is cutoff:

![Screenshot_20211103-154503_Trebuchet.png](https://user-images.githubusercontent.com/6502474/140181776-3b262dfa-e156-474c-b8ea-ee39a0547d9d.png)

But forcing a screen redraw by turning screen off and on does fix the wallpaper to fit in the new dimensions.  This is with that last apk from github CI.

**Comment by ericfont** [comment]
For example, first is a screenshot of lockscreen when viewing my phone in horizontal mode:

![Screenshot_20211103-154924.png](https://user-images.githubusercontent.com/6502474/140182457-423e6e0f-bc8b-4b62-848e-97da4a72fa9b.png)

Next screenshot after unlocking phone:

![Screenshot_20211103-154932_Trebuchet.png](https://user-images.githubusercontent.com/6502474/140182667-4826fa71-5ed9-4b10-bfed-a457fff3ff12.png)

Which seems fine...until i rotate the screen to be vertical, then i get:

![Screenshot_20211103-154503_Trebuchet.png](https://user-images.githubusercontent.com/6502474/140182792-b3fc5288-a5f6-4841-9397-f3b365adb141.png)

**Comment by forrestguice** [comment]
lol. It seems I'm gonna be a "live wallpaper lifecycle" expert by the time this thing is working right. The last 'fix' was to trigger a redraw from `onSurfaceChanged`, but it doesn't seem to be called in this case. 

I'll take a closer look. I think part of the difficulty is I don't have an actual device that allows the home screen to rotate. I don't think I would have noticed if you weren't testing.

**Comment by ericfont** [comment]
ha.  and this is interesting...if I wait 15 seconds after turning my phone vertical, then the redraw finally triggers so it looks nice.

And yeah don't feel like you have to waste too much time figuring out the quirks of android's system...I was just reporting this glitch because it was a glitch but it honestly isn't that big of a deal if it is just a matter of waiting for the redraw or forcing a redraw by turning screen on and off.

**Comment by forrestguice** [comment]
ahh. That makes sense, since the draw loop runs every 30s. ..so if happen to be looking away that long, you'll never notice ;) Maybe if the wallpaper ran more frequently (animation, several times a second) its not something that would ever come up.

**Comment by ericfont** [comment]
Yup...the 30 second draw loop makes sense.  Cause the delay happens both when going from horizontal to vertical and from vertical to horizontal...and the delay seemed to me to be a random number of seconds from 5 to 30 seconds!  Anyway, probably not even worth fixing.
